### DISCUSSION TITLE
Interactive Views not working on iOS 12.5

### DISCUSSION THREAD

**Discussion by markusgritsch** [lona-web-org/lona#334] [Q&A]
Hi,

thank you very much for your really greate Lona framework. I like it very much.

It works great across my devices, except on my iPhone 6 and iPad mini 2, where interactive Views get loaded, but only display a blank screen. I suppose there is a problem regarding the WebSocket connection.

Are these old browsers on those devices known to be not capable of this, or is there possibly a way to get them working too? I don't seem to have any compatibility issues browsing other websites. Is using WebSockets really that rare?

Kind regards,
Markus

**Comment by fscherf** [comment]
Hi @markusgritsch,

Thanks, i am glad to hear that :)

I am currently developing against very recent versions of chrome, webkit and firefox, and I dont have access to any apple devices, especially older ones. So I am not aware of problems like this, and wouldn't encounter them, at the moment.

Websockets could be the problem here. Also some dom-manipulation calls or es06 Javascript imports could be problematic.

The lona Javascript client has a mechanism that sends Javascript exceptions back to the server to log them as errors. (The websocket connection is used for that, so if you dont see that, chances are there are no connections)

You can check how many connections a view has: start a lona shell (when using a lona script, start with '--shell') and run '%lona_views'. The command has tab completion for running views. When called with an id, an list of running connections is listed.

**Reply by markusgritsch** [reply]
Not sure how to get to the shell prompt. When I start the script with `--shell`, I get `ModuleNotFoundError: No module named 'readline'`, but the webpage gets served fine. However, no `>>>` prompt as shown in the documentation anywhere.

**Reply by fscherf** [reply]
Ah! You need python with readline support to use rlpython. Do you run on windows?

**Reply by markusgritsch** [reply]
Yes, Windows. Tried running `pip install readline`, got
```
install --user readline
Collecting readline
  Downloading readline-6.2.4.1.tar.gz (2.3 MB)
     ---------------------------------------- 2.3/2.3 MB 1.8 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      error: this module is not meant to work on Windows
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
Process returned with code 1
```

**Reply by fscherf** [reply]
That's a very valuable bug-report too. I made a patch for a readline-fallback for rlpython. Could you please try to install it, and give some feedback if it fixes your problem?

```
pip uninstall rlpython
pip install git+http://github.com/fscherf/rlpython@fscherf/topic/readline-fallback
```

In the meantime: Could you check https://browserleaks.com/features on your apple-devices?

If websockets are really the problem: For other reasons, there are already ideas to add a websocket fallback to Lona. I fiddled around with sockjs, but had multiple problems with it. The plan was to create new, but similar solution, but better integrated into aiohttp and with support for continuous sessions.

https://github.com/sockjs/sockjs-client/discussions/610
https://github.com/sockjs/sockjs-client/discussions/609

**Reply by markusgritsch** [reply]
After installing your custom built `rlpython` package I get the following traceback when starting my script with `--shell`:
```
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
Traceback (most recent call last):
  File "C:\Program Files\Thonny\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Thonny\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\mgr\AppData\Roaming\Python\Python310\site-packages\rlpython\command_line.py", line 93, in <module>
    handle_command_line()
  File "C:\Users\mgr\AppData\Roaming\Python\Python310\site-packages\rlpython\command_line.py", line 30, in handle_command_line
    signal.signal(signal.SIGTSTP, signal.SIG_IGN)
AttributeError: module 'signal' has no attribute 'SIGTSTP'
```

`browserleaks.com` shows a checkmark on the `WebSockets` and `WebSockets Binary` features, so maybe it is some other missing feature. Here is the full JSON Dump:
```
{"adownload":"false","ambientlight":"false","aping":"true","appearance":"true","applicationcache":"true","areaping":"true","arrow":"true","atobbtoa":"true","audio":"true","audio_m4a":"maybe","audio_mp3":"maybe","audio_ogg":"false","audio_opus":"false","audio_wav":"maybe","audioloop":"true","backdropfilter":"true","backgroundblendmode":"true","backgroundcliptext":"true","backgroundsize":"true","batteryapi":"false","bdi":"true","beacon":"true","bgpositionshorthand":"true","bgpositionxy":"true","bgrepeatround":"true","bgrepeatspace":"true","bgsizecover":"true","blobconstructor":"true","bloburls":"true","borderimage":"true","borderradius":"true","boxdecorationbreak":"true","boxshadow":"true","boxsizing":"true","canvas":"true","canvasblending":"true","canvastext":"true","canvaswinding":"true","capture":"true","checked":"true","classlist":"true","contenteditable":"true","contextmenu":"false","cookies":"true","cors":"true","crypto":"true","cssall":"true","cssanimations":"true","csscalc":"true","csschunit":"true","csscolumns":"true","csscolumns_fill":"true","csscolumns_gap":"true","csscolumns_rule":"true","csscolumns_rulecolor":"true","csscolumns_rulestyle":"true","csscolumns_rulewidth":"true","csscolumns_span":"true","csscolumns_width":"true","cssescape":"true","cssexunit":"true","cssfilters":"true","cssgradients":"true","cssgrid":"true","cssgridlegacy":"false","cssinvalid":"true","cssmask":"true","csspointerevents":"true","csspositionsticky":"true","csspseudoanimations":"true","csspseudotransitions":"true","cssreflections":"true","cssremunit":"true","cssresize":"true","cssscrollbar":"true","csstransforms":"true","csstransforms3d":"true","csstransformslevel2":"false","csstransitions":"true","cssvalid":"true","cssvhunit":"true","cssvmaxunit":"false","cssvminunit":"true","cssvwunit":"true","cubicbezierrange":"true","customelements":"true","customevent":"true","customproperties":"true","customprotocolhandler":"false","dart":"false","datachannel":"true","dataset":"true","dataview":"true","details":"true","devicemotion":"true","deviceorientation":"true","display-runin":"false","displaytable":"true","documentfragment":"true","effectivetype":"false","ellipsis":"true","emoji":"true","es5":"true","es5array":"true","es5date":"true","es5function":"true","es5object":"true","es5string":"true","es5syntax":"true","es5undefined":"true","es6array":"true","es6collections":"true","es6math":"true","es6number":"true","es6object":"true","es6string":"true","es7array":"true","es8object":"true","eventlistener":"true","eventsource":"true","fetch":"true","fileinput":"true","fileinputdirectory":"true","filereader":"true","filesystem":"false","flexbox":"true","flexboxlegacy":"true","flexboxtweener":"false","flexgap":"false","flexwrap":"true","focusvisible":"false","focuswithin":"true","fontdisplay":"true","fontface":"true","forcetouch":"false","formattribute":"true","formvalidation":"true","framed":"false","fullscreen":"false","gamepads":"true","generatedcontent":"true","generators":"true","geolocation":"true","getrandomvalues":"true","getusermedia":"false","hairline":"true","hashchange":"true","hidden":"true","history":"true","hovermq":"false","hsla":"true","htmlimports":"false","ie8compat":"false","imgcrossorigin":"true","inlinesvg":"true","input_autocomplete":"true","input_autofocus":"true","input_list":"true","input_max":"true","input_min":"true","input_multiple":"true","input_pattern":"true","input_placeholder":"true","input_required":"true","input_step":"true","inputformaction":"true","inputformenctype":"true","inputformmethod":"true","inputformnovalidate":"true","inputformtarget":"true","inputsearchevent":"true","inputtypes_color":"true","inputtypes_date":"true","inputtypes_datetime":"false","inputtypes_datetime-local":"true","inputtypes_email":"true","inputtypes_month":"true","inputtypes_number":"true","inputtypes_range":"true","inputtypes_search":"true","inputtypes_tel":"true","inputtypes_time":"true","inputtypes_url":"true","inputtypes_week":"true","intersectionobserver":"true","intl":"true","json":"true","lastchild":"true","lazyloading":"false","ligatures":"true","localstorage":"true","lowbandwidth":"false","lowbattery":"false","matchmedia":"true","mathml":"true","mediaqueries":"true","mediasource":"false","mediastream":"false","messagechannel":"true","meter":"true","microdata":"false","multiplebgs":"true","mutationobserver":"true","notification":"false","nthchild":"true","objectfit":"true","olreversed":"true","oninput":"true","opacity":"true","outputelem":"true","overflowscrolling":"true","pagevisibility":"true","passiveeventlisteners":"true","peerconnection":"true","performance":"true","picture":"true","pointerevents":"false","pointerlock":"false","pointermq":"true","postmessage":"true","postmessage_structuredclones":"true","prefetch":"false","preserve3d":"true","progressbar":"true","promises":"true","proxy":"true","publickeycredential":"false","queryselector":"true","quotamanagement":"false","regions":"false","requestanimationframe":"true","requestautocomplete":"false","resizeobserver":"false","restdestructuringarray":"true","restdestructuringobject":"true","restparameters":"true","rgba":"true","ruby":"true","sandbox":"true","scriptasync":"true","scriptdefer":"true","scrollsnappoints":"true","scrolltooptions":"true","seamless":"false","serviceworker":"false","sessionstorage":"true","shadowroot":"true","shadowrootlegacy":"false","shapes":"true","sharedworkers":"false","siblinggeneral":"false","smil":"true","speechrecognition":"false","speechsynthesis":"true","spreadarray":"true","spreadobject":"true","srcdoc":"true","srcset":"true","strictmode":"true","stringtemplate":"true","stylescoped":"false","subpixelfont":"true","supports":"true","svg":"true","svgasimg":"true","svgclippaths":"true","svgfilters":"true","svgforeignobject":"true","target":"true","template":"true","textalignlast":"false","textareamaxlength":"true","textdecoder":"true","textdecoration":"true","textdecoration_color":"true","textdecoration_line":"true","textdecoration_skip":"true","textdecoration_skipink":"false","textdecoration_style":"true","textencoder":"true","textshadow":"true","texttrackapi":"true","time":"false","todataurljpeg":"true","todataurlpng":"true","todataurlwebp":"false","track":"true","typedarrays":"true","unicoderange":"true","unknownelements":"true","urlparser":"true","urlsearchparams":"true","userdata":"false","userselect":"true","variablefonts":"true","vibrate":"false","video":"true","video_h264":"probably","video_hls":"probably","video_ogg":"false","video_vp9":"false","video_webm":"false","videocrossorigin":"true","videoloop":"true","videopreload":"true","vml":"false","webanimations":"false","webaudio":"true","webintents":"false","websockets":"true","websocketsbinary":"true","websqldatabase":"true","webworkers":"true","willchange":"true","wrapflow":"false","xdomainrequest":"false","xhr2":"true","xhrresponsetype":"true","xhrresponsetypearraybuffer":"true","xhrresponsetypeblob":"true","xhrresponsetypedocument":"true","xhrresponsetypejson":"true","xhrresponsetypetext":"true"}```

**Reply by markusgritsch** [reply]
Also, I cannot stop the server using `CTRL+C` when using the new `rlpython`, but have to use the Windows task manager to kill the python.exe process.

**Reply by fscherf** [reply]
The rlpython issue seems to be more complicated. Thanks for trying, i will add windows testing to rlpython to resolve this issue.

I looked over the browserleaks dump, and saw nothing obvious missing. Maybe we should check first if a simpler websocket app works.
I can create a simple aiohttp+html+js snippet for you til the end of the weekend, if you need. If that works, we know the problem is in lona.

Do you think there is a way to get an actual error message from the browser?

**Reply by markusgritsch** [reply]
Unfortunately I have no idea, how to get some error message from the browser on my iPhone.

Regarding the simple websocket app: The simple chat application at the bottom of https://javascript.info/websocket works fine on my iPhone. The online version from the website and also when served locally from my PC.

**Reply by fscherf** [reply]
Ah ok, thats a start. Next thing i would suspect are es06 modules, lona uses: https://www.w3schools.com/js/js_modules.asp

**Reply by markusgritsch** [reply]
The `Try it yourself` buttons on that page all give the same output on my iPhone as on my PC.

**Reply by fscherf** [reply]
Hm then i am running out of ideas now. Lona has two different client implementations since 1.12. Could you try client2 by setting `CLIENT_VERSION=2` in the settings?

**Reply by markusgritsch** [reply]
I tried running my script with the command line option ` -o CLIENT_VERSION=2`, but still not succeeding on my iPhone.

**Reply by markusgritsch** [reply]
I tried installing Safari for Windows following the download link form this page: https://www.intowindows.com/how-to-install-safari-browser-on-windows-10/

It is quite old, and also does not show any content. I don't know, if the reason is the same, but at least it gave an error message:
![grafik](https://user-images.githubusercontent.com/2684183/218281372-1aecf83e-9a5b-48f1-8ed9-c309efad27a7.png)

But maybe this browser is so old, that it has nothing to do with the reason it does not work on my iPhone.

**Reply by fscherf** [reply]
**client version:** `-o CLIENT_VERSION=2` will have no effect because `-o` (lower case o) does settings pre overrides. That means the client version gets set before all other settings get loaded, including `lona.default_settings`. `lona.default_settings` sets the client version to 1, so what you need to do, if you want to set it via command line, is a post override `-O` (uppercase o), which gets set after all settings files are loaded.

**safari:** That could be the problem. `Lona` is defined in line 12, and in line 24 it is not accessible. The problem could be that line 23 does not use an arrow function, which would inherit the parents scope. I assume you use the default frontend template in Lona. Could you try this patched one?

```
pip uninstall lona
pip install git+http://github.com/lona-web-org/lona@fscherf/topic/safari-debugging
 ```

**Reply by markusgritsch** [reply]
The arrow function does not work either:
![grafik](https://user-images.githubusercontent.com/2684183/218297044-fe7afd32-ef61-4587-8933-77cf75ed7424.png)

Already tried `-O` with my iPhone, also no success.

**Reply by markusgritsch** [reply]
I think it's of no big help trying to get lona running on this unmaintained 10+ years old Windows Safari browser. There are probably lots of reasons, it won't work on this outdated version. The simple WebSocket chat example from above at https://javascript.info/websocket (which worked fine on my iPhone) is also not working.

**Reply by markusgritsch** [reply]
Great news: I am now able to debug directly on my iPhone using this software: https://inspect.dev/ (13 days left 😉)

Here is the error message using the arrow function:
![grafik](https://user-images.githubusercontent.com/2684183/218300325-2ebf691f-fee7-4cb0-a594-8e8f5f249d88.png)

It's the same as with the original function() syntax:
![grafik](https://user-images.githubusercontent.com/2684183/218300388-10412ce5-d580-4364-af9d-90eac04030d5.png)

**Reply by markusgritsch** [reply]
Do you have an idea, where this error message comes from:
`(unknown) SyntaxError: Unexpected token '='. Expected an opening '(' before a method's parameter list.`

![grafik](https://user-images.githubusercontent.com/2684183/218301108-597237cb-49d9-43ca-b3a9-587480d1d10d.png)

**Reply by fscherf** [reply]
Ouh nice! :) I researched browserstack.com in the meantime, but it's not cheap.

I found a similar issue on stackoverflow: https://stackoverflow.com/questions/60026651/safari-unexpected-token-expected-an-opening-before-a-methods-paramet

Lona uses puplic field declarations in fact! So that could be the problem. I added a patch to my safari debugging branch. Could you test it again? You need to use `-O CLIENT_VERSION=2` because i only patched the new client.

**Reply by markusgritsch** [reply]
Yay! It worked! Here's a screenshot from my Phone:

![IMG_8423](https://user-images.githubusercontent.com/2684183/218305284-3dccc996-af1e-4d2e-b6e4-8ff5ee6c2d57.PNG)

**Reply by markusgritsch** [reply]
I don't know if this is related to your modification in CLIENT_VERSION 2, but the following simple Bootstrap 5 example does only work with the old client version:
```python
from lona_bootstrap_5 import (
    Switch,
    BorderPrimarySpinner,
)
from lona import LonaApp, LonaView
from lona.html import H1, HTML, P, Div

from lona_bootstrap_5.nodes import BorderInfoSpinner, PrimaryBadge

app = LonaApp(__file__)

app.add_static_file('lona/style.css', """
#lona {
    font-family: sans-serif;
    color: #333333;
    max-width: 60em;
    margin: 0 auto;
}
""")

@app.route('/')
class MyView(LonaView):
    def handle_request(self, request):
        self.spinner = BorderPrimarySpinner('Test')

        html = HTML(
            H1('Bootstrap 5'),

            Switch(value=True, handle_change=self.handle_switch_change),

            self.spinner,
        )
        return html

    def handle_switch_change(self, event):
        self.spinner.style['visibility'] = 'visible' if event.data else 'hidden'

app.run(host='0.0.0.0', port=8080)
```
With CLIENT_VERION 2 it gives a `500 Internal error`

**Reply by fscherf** [reply]
Yay! :) that was expected, because lona-bootstrap-5 has no support for client 2 yet. I will port this fix client 1, this evening, and notify you for another round of testing.

**Reply by markusgritsch** [reply]
`lona.js` seems to be identical in `client` and `client2`, so I copied the fixed one from `client2` to `client` and now also the Bootstrap 5 example works on my iPhone 6 👍.

**Reply by fscherf** [reply]
@markusgritsch: yay! I created a PR, that patches both clients: https://github.com/lona-web-org/lona/pull/335. Could you verify that it resolves the issue?

Thanks a lot for your work and debugging! You helped me more than you know(!)
Feel free to add yourself to the contributors file, and send me an email if you want some stickers.
